### ISSUE TITLE
Calling GetObjectCommand with Range returns undefined in response.ContentRange

### ISSUE THREAD

**Issue by rtarling** [aws/aws-sdk-js-v3#5650]
### Checkboxes for prior research

- [X] I've gone through [Developer Guide](https://docs.aws.amazon.com/sdk-for-javascript/v3/developer-guide) and [API reference](https://docs.aws.amazon.com/AWSJavaScriptSDK/v3/latest)
- [X] I've checked [AWS Forums](https://forums.aws.amazon.com) and [StackOverflow](https://stackoverflow.com/questions/tagged/aws-sdk-js).
- [X] I've searched for [previous similar issues](https://github.com/aws/aws-sdk-js-v3/issues) and didn't find any solution.

### Describe the bug

Hi

The GetObjectCommand is not returning ContentRange to the client in the response object when passing in a Range, and running in the browser/webview.

I am running the latest version of the SDK and Chrome/Android Webview.

Hope you can help as I am stuck.

Thanks
Roland

### SDK version number

@aws-sdk/client-s3"  "version": "3.485.0"

### Which JavaScript Runtime is this issue in?

Browser

### Details of the browser/Node.js/ReactNative version

Chrome Version 120.0.6099.200 (Official Build) (64-bit) + Android Webview

### Reproduction Steps

I am trying to implement a chunked *download* of files using the AWS example: 

https://docs.aws.amazon.com/AmazonS3/latest/userguide/example_s3_Scenario_UsingLargeFiles_section.html

*** Note that I am doing this from a browser/webview, not from a lambda or CLI. ***

The example expects:

const command = new GetObjectCommand({
      Bucket: bucket,
      Key: key,
      Range: `bytes=${start}-${end}`,
});

s3Client.send(command);

### Observed Behavior

The command does return the correct number of bytes, however the response.ContentRange is always undefined.  When I look at the network output it shows that the content range is being returned in the response headers - it is just not being mapped to the response object.

I see there was a similar issue in this ticket but it was never followed up by the raiser.  https://github.com/aws/aws-sdk-js-v3/issues/5004

### Expected Behavior

The response.ContentRange should return the bytes downloaded/remaining so that the next download chunk can be calculated ands retrieved.

### Additional Information/Context

I have asked AWS support to look into this and they can reproduce the error but don't know how to fix it.  I suspect they will also contact you at some point.

**Comment by RanVaknin** [comment]
Hi @rtarling ,

I see that you have opened a support ticket internally, I will gladly interact with you here but I will also inform the support engineer assigned to your ticket of any new information.

From their investigation they were not able to reproduce the reported behavior where contentRange is missing when requested. I did my own investigation and also was not able to reproduce. In my example the contentRange is serialized and printed correctly:

```js
import { S3Client, GetObjectCommand } from "@aws-sdk/client-s3";

const s3 = new S3Client({ region: "us-east-1" });

async function foo(){
  const params = {
    Bucket: "foo-bucket",
    Key: "sample.xlsx",
    Range: "bytes=1-2"
  };
  
  try {
    const res = await s3.send(new GetObjectCommand(params));
    console.log(res.ContentRange)
  } catch (error) {
    console.error(error)
  }
};
foo() 
// bytes 1-2/4922
```

Are you using a proxy that might be rewriting response headers? Maybe try a different browser?

You can check the data before deserialization which should be the raw http response by doing the following:

```js
 s3Client.middlewareStack.add(next => async (args) => {
 console.log("Raw Request:", args.request)
 const response = await next(args);
 console.log("Raw Response:", response);
 return next(args);
}, {step: 'finalizeRequest'})

const response = await s3Client.send(
        new GetObjectCommand({
          Bucket: import.meta.env.VITE_BUCKET_NAME,
          Key: sourceFileName,
          Range: "bytes=1-2",
        })
      );
// rest of your code.
```

If you can inspect the raw request and response provided by this middleware and share it with us, it will tell us the following:

- If the data printed from the middleware is missing Content-Range then the browser alters the response structure (this 
might be related to Vite, without further context It's hard to know why this is happening.
- If the Content-Range is present in the middleware log but not in their response.ContentRange. This might point to a potential deserialization issue with the SDK. If this is the case we would need to see the raw logs to further investigate.

All the best,
Ran~

**Comment by rtarling** [comment]
Hi Ran

Many thanks for the quick turnaround – appreciate that.  Attached are the 3 logs you requested.  Hope you find the problem in them.

Kind Regards
Roland

**Comment by rtarling** [comment]
Hi Ran

Any luck on this?

Thanks
Roland

From: Roland Tarling
Sent: Sunday, 14 January 2024 13:49
To: aws/aws-sdk-js-v3 ***@***.***>
Subject: RE: [aws/aws-sdk-js-v3] Calling GetObjectCommand with Range returns undefined in response.ContentRange (Issue #5650)

Hi Ran

Many thanks for the quick turnaround – appreciate that.  Attached are the 3 logs you requested.  Hope you find the problem in them.

Kind Regards
Roland

**Comment by rtarling** [comment]
Hi @RanVaknin 

I added the 3 files you requested 2 weeks ago but no response yet.  Do you need anything more from me?

Thanks
Roland

**Comment by rtarling** [comment]
@RanVaknin here are the logs again.
[response.log](https://github.com/aws/aws-sdk-js-v3/files/14060717/response.log)
[rawresponse.log](https://github.com/aws/aws-sdk-js-v3/files/14060718/rawresponse.log)
[rawrequest.log](https://github.com/aws/aws-sdk-js-v3/files/14060719/rawrequest.log)

**Comment by RanVaknin** [comment]
Hi @rtarling ,

Thank you for your patience, and apologies for the long wait times. 
I looked through the logs provided. It seems like the raw response you are seeing does not contain that header at all, therefore the SDK does not have anything to deserialize. 

This will have to get investigated with the S3 team itself since they are not sending back the expected header in the response.
 I will contact your technical account manager and have them put you in touch with the S3 team internally.
 
 Thanks again,
 Ran~

**Comment by rtarling** [comment]
Hi Ran

Thanks for getting back to me.

This is a bit strange isn’t it that S3 isn’t returning the header as that is probably the most used and tested of all your products...

I don’t have a technical account manager.  I only opened a ticket on the AWS help center.  The S3 team can contact me via the ticket or email me directly.

Thanks
Roland

**Comment by sbelbin** [comment]
I can state that I've been experiencing the same issue as reported when running the sample code in Chrome Browser, such that the JavaScript v3 AWS API that the response isn't providing the **ContentRange**.

However, we can observe in attached image below taken from the browser's Network traffic log that HTTP response that its  clearly returned.

_The code used is that of from https://docs.aws.amazon.com/sdk-for-javascript/v3/developer-guide/javascript_s3_code_examples.html_

The response from sending the **GetObjectCommand** doesn't contain a **ContentRange** field. Therefore, the total size of the underlying object in the S3 bucket cannot be determined as defined in the source code example.

**@aws-sdk/client-s3** - 3.537.0
This was done using an fresh npm install that I've done today.

I suggest that someone take a look at because there is a high probablity that many developers have implemented such a solution into their application based on the sample that AWS provided.

![aws-getobject-content-ranges-woes](https://github.com/aws/aws-sdk-js-v3/assets/159043895/8d705c9c-f784-484d-9772-ef1726f03089)

**Comment by rtarling** [comment]
@sbelbin I finally got an explanation via AWS Support...after weeks.

You need to update your bucket CORS policy and add "content-range" header to the ExposeHeader list.  Not obvious at all...

A useful link about it is here:

https://kikobeats.com/s3-cors/

Hope this helps!
Roland

**Comment by sbelbin** [comment]
> @sbelbin I finally got an explanation via AWS Support...after weeks.
> 
> You need to update your bucket CORS policy and add "content-range" header to the ExposeHeader list. Not obvious at all...
> 
> A useful link about it is here:
> 
> https://kikobeats.com/s3-cors/
> 
> Hope this helps! Roland

I appreciate the insight and I'll make the corresponding changes. However, as I've stated this is based from AWS sample code, therefore a mention should be made to the sample or it should be revised accordingly.

In addition, the response from the AWS server as I've demonstrated in the screen capture that the **Content-Range** was returned from the AWS S3 service. Thus, the the AWS S3Client took extra steps to omit that information as part of fields its **GetObjectCommandOutput**.

Thanks,
Steven

**FYI:**
Just to share, the workaround that I was using by making a **GetObjectCommand** request for the total size of the underlying object then followed by changes to the sample code as to fetch in segments and using that initial value to track the number of remaining bytes.

**Comment by rtarling** [comment]
Hi Steve

In my experience the AWS samples often don’t work, are out of date, or too simplistic to be useful.  I find GPT is sometimes more useful than the AWS docs themselves, as every now and then it reveals a nugget of information that I could have spend days looking for on AWS...

Anyway, I can only say I made the change and it worked for me:

        "ExposeHeaders": [
            "x-amz-server-side-encryption",
            "x-amz-request-id",
            "x-amz-id-2",
            "Content-Length",
            "Content-Range",
            "ETag"
        ],

Thanks
Roland

**Comment by RanVaknin** [comment]
Hi All, 

I did mention in my [previous comment](https://github.com/aws/aws-sdk-js-v3/issues/5650#issuecomment-1964707332) that you might want to configure CORS if its not enabled on your bucket. 

Regarding the example from the https://docs.aws.amazon.com/sdk-for-javascript/v3/developer-guide/javascript_s3_code_examples.html

This examples doc list is a basic list of operations and would not include any combination of headers / configurations that S3 offers as it will be too extensive to cover. It will only show you how to use the SDK, but the CORS related issue that was reported is not specific to the SDK therefore the documentation on how to circumvent this scenario would be on the [S3 docs](https://docs.aws.amazon.com/sdk-for-javascript/v3/developer-guide/cors.html) [themselves](https://docs.aws.amazon.com/AmazonS3/latest/userguide/cors.html) and not the SDK docs.

I'm glad to hear you are both unblocked. I'm going to go ahead and close this issue. If this still persists please feel free to create a new issue.

All the best,
Ran~

**Comment by sbelbin** [comment]
I appreciate your follow-up on this matter and I can confirm that explicitly adding the aforementioned ‘Content-Range’ to the CORS settings resolves the matter. Thus, feel free to close the issue with a recommendation as you’ve stated to clarify in the developer documentation (demonstration).

A technical concern is that as demonstrated in network payload the HTTP response contains the ‘Contents-Range’ regardless of the CORS settings. Hence developer(s) choose not to unless that setting explicitly defined in the CORS settings.
If it were missing the network payload, then reviewing the CORS settings would 

Thus, IMHO the API isn’t orthogonal and is misleading. This fails the "pitfall of success” which all of us in the developer community should strive for. I’ve added it to my bucket list to see if I do a fix in the code base, I’ll proceed to submit a code fix for review.
